rust/metering/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+    utils::{generate_compile_error, process_token_stream},
+};
+
+mod annotations;
+mod attributes;
+mod errors;
+mod events;
+mod fields;
+mod mutators;
+mod utils;
+
+/// This is the only user-facing export of `chroma_metering`. It is responsible for registering attributes and
+/// events by producing the code necessary to allow users to interact with the metering library.
+#[proc_macro]
+pub fn initialize_metering(raw_token_stream: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let token_stream = TokenStream::from(raw_token_stream);
+
+    let (attributes, events) = match process_token_stream(&token_stream) {
+        Ok(result) => result,
+        Err(error) => return generate_compile_error(&error.to_string()),
+    };
+
+    let noop_mutator_definition_token_streams: Vec<TokenStream> = attributes
+        .iter()
+        .map(generate_noop_mutator_definition_token_stream)
+        .collect();
+
+    let attribute_definition_token_streams: Vec<TokenStream> = attributes
+        .iter()
+        .map(generate_attribute_definition_token_stream)
+        .collect();
+
+    let event_definition_token_streams: Vec<TokenStream> = events
+        .iter()
+        .map(generate_event_definition_token_stream)
+        .collect();
+
+    proc_macro::TokenStream::from(quote! {
+        /// The primary trait used in the metering library that contains no-op mutators for every attribute.
+        pub trait MeteringEvent: std::fmt::Debug + std::any::Any + Send + 'static {
+            #( #noop_mutator_definition_token_streams )*
+        }
+
+        #( #attribute_definition_token_streams )*
+
+        #( #event_definition_token_streams )*
+
+        /// The default receiver registered in the library.
+        #[derive(Clone, std::fmt::Debug)]
+
+        pub struct __DefaultReceiver;
+
+        /// The default receiver simply prints out the metering events submitted to it.
+        #[async_trait::async_trait]
+        impl chroma_system::ReceiverForMessage<Box<dyn MeteringEvent>>
+            for __DefaultReceiver
+        {
+            async fn send(
+                &self,
+                message: Box<dyn MeteringEvent>,
+                tracing_context: Option<tracing::Span>,
+            ) -> Result<(), chroma_system::ChannelError> {
+                if let Some(span) = tracing_context {
+                    println!("[metering] span={:?} event={:?}", span, message);
+                } else {
+                    println!("[metering] event={:?}", message);
+                }
+                Ok(())
+            }
+        }
+
+        /// The storage slot for the registered receiver.
+        static RECEIVER: once_cell::sync::Lazy<
+            parking_lot::Mutex<Box<dyn chroma_system::ReceiverForMessage<Box<dyn MeteringEvent>>>>,
+        > = once_cell::sync::Lazy::new(|| parking_lot::Mutex::new(Box::new(__DefaultReceiver)));
+
+        /// Allows library users to register their own receivers.
+        pub fn register_receiver(
```

Comment:
I think this name is confusing, it implies to the caller multiple potential recievers can exist, however, only one can. The old library called this `init_receiver` and I think that is a better name.